### ISSUE TITLE
Application Service support tracking

### ISSUE THREAD

**Issue by johannescpk** [matrix-org/matrix-rust-sdk#228]
The matrix-sdk-appservice crate should be considered experimental until this issue gets closed

Docs: https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk_appservice/index.html

### Bugs

- [ ] regex matching on registration namespaces chokes on escape sequences which makes it incompatible to how synapse handles it (e.g. `'@uwu:owo\.tld'`)
- [x] Running the appservice on a sub-path will not work correctly as ruma only parses URLs correctly that start with `/_matrix` as root (https://github.com/matrix-org/matrix-rust-sdk/pull/304)
- [x] `register` not properly working (https://github.com/matrix-org/matrix-rust-sdk/pull/254)
- [x] Registering [scope `/` for the fallback routes](https://github.com/matrix-org/matrix-rust-sdk/blob/ffea84b64aa172b9b153a76588f4f609bf15c441/matrix_sdk_appservice/src/actix.rs#L55) breaks consumer routes that are added after adding `actix_service` (https://github.com/matrix-org/matrix-rust-sdk/pull/254)
- [x] `if let Ok(client) = self.appservice.client(None).await {};` in EventHandler doesn't seem to work (https://github.com/matrix-org/matrix-rust-sdk/pull/254)

### Hard
- [x] Correctly recognize room membership based on namespaces (https://github.com/matrix-org/matrix-rust-sdk/issues/507, https://github.com/matrix-org/matrix-rust-sdk/pull/710, https://github.com/matrix-org/matrix-rust-sdk/pull/747)
- [ ] Transaction tokens must be handled differently than the sync response token. (https://github.com/matrix-org/matrix-rust-sdk/pull/853)
- [ ] Properly sync room state
  - Encountering a room the first time when processing a transaction should block further transaction processing and sync the state events for the room first. PR: https://github.com/matrix-org/matrix-rust-sdk/pull/303
- [ ] E2EE
  - ~~Blocked by waiting for the implementation work around [MSC3202](https://github.com/matrix-org/matrix-doc/pull/3202). There are workarounds but the added complexity just isn't worth.~~ Ready to be picked up: https://github.com/matrix-org/matrix-spec-proposals/pull/3202#issuecomment-1185756169

### Medium
- [x] Expose user/room queries from the homeserver (https://matrix.org/docs/spec/application_service/r0.1.2#querying)
- [x] Persist info about registered virtual users in the arbitrary SDK storage (#315)
- [x] Basic test that checks whether ClientConfig & RequestConfig are correctly used
- [x] Singleton for `Appservice::client()` (https://github.com/matrix-org/matrix-rust-sdk/pull/249)
- [x] Expose `ClientConfig` (https://github.com/matrix-org/matrix-rust-sdk/pull/249)
- [x] Add `warp` support (https://github.com/matrix-org/matrix-rust-sdk/pull/254)
  - [x] Fully working support is blocked behind https://github.com/seanmonstar/warp/pull/861, workaround is 
        ```
        [patch.crates-io]
        warp = { version = "0.3", git = "https://github.com/johannescpk/warp", rev = "4ead0f", default-features = false, features = ["multipart", "websocket"] }
        ```
- [x] Refactor actix example to use `?` instead of `unwrap`

### Easy
- [x] Don't process transactions twice (https://github.com/matrix-org/matrix-rust-sdk/pull/560)
- [x] Proper ~~re-export and~~ usage of ruma appservice api instead of hidden/inlined re-export (related: https://github.com/matrix-org/matrix-rust-sdk/issues/152)
  - [x] Using re-exported `ruma` from a consumer doesn't completely work since it breaks the `EventContent` macro
- [x] Switch warp dependency to crates.io `version`, blocked behind new warp release (after 0.3.1)
- [x] docs: Explain that incoming transactions are handled by the main user client (https://github.com/matrix-org/matrix-rust-sdk/pull/303)
- [x] Allow to drop or re-initialize cached singleton clients. Done via `virtual_users` returning the dashmap.
- [x] Mention in the docs that the appservice is basically a wrapper around the sdk `Client`
- [x] Remove outdated `serde_yaml` dependency from matrix-sdk crate
- [x] Fix typo: `try_into_http_request_with_identy_assertion` -> `try_into_http_request_with_identity_assertion` :D
- [x] Rename `client_with_localpart` to `client_with_identity_assertion` (https://github.com/matrix-org/matrix-rust-sdk/pull/231)
- [x] Move `get_host_and_port_from_registration` into `AppserviceRegistration::get_host_and_port` (https://github.com/matrix-org/matrix-rust-sdk/pull/231)
- [x] Check if it makes sense to introduce an `InnerAppservice` for `Arc`-purposes, or single `Arc`s for member fields (from https://github.com/matrix-org/matrix-rust-sdk/pull/204#discussion_r625759188)
- [x] Initial version (https://github.com/matrix-org/matrix-rust-sdk/pull/204)

### Refactoring
- [x] matrix_sdk_test: use `SyncStateEvent::into_full_event` instead of `value_with_room_id`
- [ ] Try intermediate conversion from `http::Request` instead (see https://github.com/matrix-org/matrix-rust-sdk/pull/254#discussion_r646015380)
- [x] Consider renaming `Appservice` to `AppService` (https://github.com/matrix-org/matrix-rust-sdk/pull/272)
- [x] `localpart`-unwrap in `Appservice::client` and `Appservice::client_with_config` should be deduplicated and re-use get_cached_client instead, but was hitting some Send + Sync roadblock
- [x] Using `appservice` in the actix examples probably wants a mutex instead of cloning, or maybe allow `&mut self` event handler methods
- [x] Make `localpart` in `client_with_config()` a non-Option, since the main appservice user config can't get overwritten anyway

### Optional / Nice to have
- [x] Consider removing actix (https://github.com/matrix-org/matrix-rust-sdk/pull/290)
  - Reasoning: actix does a rather specific tokio runtime setup that leads to being able to easily block threads
- [ ] Think about appservice-FFI scenarios: would spawning the webserver in a thread already help? how to sanely FFI events / event handler?
- [x] Support hyper: maybe merge in from [matrix-appservice-rs](https://github.com/lieuwex/matrix-appservice-rs)? /cc @lieuwex
   - hyper is supported by using the warp filter as tower service
- [ ] Consider supporting constructing clients based on server_name & well-known
  - Related PR that introduces well-known support in the Client: https://github.com/matrix-org/matrix-rust-sdk/pull/233

### Constant effort

- Tracing
- Test-Coverage

### Other Application Service Implementations

(todo: feature matrix table?)

- https://github.com/matrix-org/matrix-appservice-node (JS, No E2EE OOTB)
- https://github.com/turt2live/matrix-bot-sdk (JS, Basic E2EE Support (https://github.com/turt2live/matrix-bot-sdk/pull/142))
- https://github.com/tulir/mautrix-go (Go, No E2EE OOTB)
  - https://github.com/tulir/mautrix-whatsapp (Go, E2EE OOTB)
- https://github.com/tulir/mautrix-python (Python, E2EE OOTB)
- https://github.com/lieuwex/matrix-appservice-rs (Rust, Standalone Client, No E2EE OOTB)

### Bridge projects

- https://github.com/matrix-org/matrix-appservice-bridge (JS, Based on matrix-appservice and matrix-js-sdk)
- https://github.com/Sorunome/mx-puppet-bridge (JS, Based on matrix-bot-sdk)
- https://github.com/matrix-hacks/matrix-puppet-bridge (JS, Based on matrix-appservice-bridge)
- https://github.com/tulir/mautrix-python (Python)

### Related projects
- Rust: https://github.com/matrix-org/matrix-ircd creates a matrix client per IRC user connection

**Comment by MTRNord** [comment]
Additional to the membership Bug is that invites only go to the invite handler when they come in via the invite branch of the timeline. This doesnt seem to be an easy fix as currently a) all appservice Events come via the join Part (due to ruma) and b) the Event handler Part where this gets relayed is struct typed. So the joined room struct cant be passed along to make this work from the join Part.

**Comment by johannescpk** [comment]
@MTRNord This should be fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/254 which also includes a bugfix due to improved test coverage. The issue being that identity assert and auth forcing didn't work as expected any longer.

**Comment by johannescpk** [comment]
@MTRNord Also, about `invite` vs `join` events, please see https://docs.ruma.io/ruma/api/appservice/event/push_events/v1/struct.IncomingRequest.html#method.try_into_sync_response for the rationale. Added that text to the `set_event_handler` docs as well now. Hope that makes sense. If you have different ideas about it, please let me know.

**Comment by agraven** [comment]
I think item 4 in "Nice to have" can be checked since the linked PR has been merged

**Comment by johannescpk** [comment]
@agraven Clarified that item's description. The linked PR was only about introducing it on the `Client`, I think it needs more work to support it in the appservice crate

**Comment by Hywan** [comment]
I'm closing this issue since https://github.com/matrix-org/matrix-rust-sdk/pull/2509 has removed `matrix-sdk-appservice`.
